### DISCUSSION TITLE
Light Source Aliasing

### DISCUSSION THREAD

**Discussion by Xaldew** [RayTracing/raytracing.github.io#964] [Q&A]
Hello!

This is something that's bugged me a little bit while doing the later books in this series: For some reason all area light sources appear to have a significant amount of aliasing / stair-casing artifacts and it is especially apparent in the final scene image in the "Next Week" book, or in my own example scene with a disk-shaped light-source:

![img-2 22-book2-final](https://user-images.githubusercontent.com/1394310/143457461-0f6e78cf-50cc-4a2d-ac97-65e554bbd394.jpeg)
![cylinders](https://user-images.githubusercontent.com/1394310/143457496-ffe58382-cf67-47e6-a507-4c08109654c2.png)

Is this a known problem with some obvious solution? As far as I'm aware, increasing pixel sample count (with sufficient in-pixel randomness) should have reduced this kind of error, but this artifact have persisted thus far with no obvious change throughout the book series

Best regards

**Comment by hollasch** [comment]
Man, that's a really good question; it never occurred to me.

**Reply by Xaldew** [reply]
I spent a bit of time looking into this the other day and my conclusion is that this appears to be "working as intended". My understanding is that multiple pixel samples *are* contributing to the pixels to remove aliasing, but since the light is so much brighter than the surroundings (which are almost 0 in the examples above), hitting even a fraction of the light immediately overpowers the surrounding pixels, yielding this stair-stepping effect.

In the real world, this would manifest itself as "bloom", but here we keep adding to the dynamic range, retaining more information about each pixel, which ultimately look a bit different from what we would expect.

As an example, the same "issue" also appear in traditional renderers, e.g., PBRT rendering the traditional Multiple Importance Sampling scene from Veach and Guibas:

![veach-mis](https://user-images.githubusercontent.com/1394310/196433242-2d2affec-08a1-4851-90d7-f23f85286f42.png)

Now, if one want to *remove* this, I believe the traditional approach is to use [tone-mapping](https://64.github.io/tonemapping/), possibly in combination with some blurring to simulate Bloom, see e.g., [LearnOpenGL](https://learnopengl.com/Advanced-Lighting/Bloom). As an example, adding the extended Reinhard tone-mapping yields the following cylinder scene from my previous post:

![cylinders](https://user-images.githubusercontent.com/1394310/196439037-f2948eed-81f1-4d56-9f48-2abc696521ec.png)

Which looks a *little* better at least.

I'm not sure if this is mentioned anywhere already in the books, if not, it is probably a good idea to add a footnote about it, especially close to the final image in Book 2 where the "artifact" is pretty obvious
